utils/service_discovery/sd_backend.py

```diff
@@ -0,0 +1,163 @@
+# std
+import logging
+import re
+import simplejson as json
+
+# project
+from utils.config_stores import ConfigStore
+from utils.dockerutil import get_client as get_docker_client
+
+log = logging.getLogger(__name__)
+
+
+class ServiceDiscoveryBackend(object):
+    """Singleton for service discovery backends"""
+    _instance = None
+
+    def __new__(cls, *args, **kwargs):
+        if cls._instance is None:
+            agentConfig = kwargs.get('agentConfig', {})
+            if agentConfig.get('service_discovery_backend') == 'docker':
+                cls._instance = object.__new__(SDDockerBackend, agentConfig)
+            else:
+                log.error("Service discovery backend not supported. This feature won't be enabled")
+                return
+        return cls._instance
+
+    def __init__(self, agentConfig=None):
+        self.PLACEHOLDER_REGEX = re.compile(r'%%.+?%%')
+        self.agentConfig = agentConfig
+
+    def get_configs(self):
+        """Get the config for all docker containers running on the host."""
+        raise NotImplementedError()
+
+    def _render_template(self, init_config_tpl, instance_tpl, variables):
+        """Replace placeholders in a template with the proper values.
+           Return a list made of `init_config` and `instances`."""
+        config = [init_config_tpl, instance_tpl]
+        for tpl in config:
+            for key in tpl:
+                for var in self.PLACEHOLDER_REGEX.findall(str(tpl[key])):
+                    if var.strip('%') in variables and variables[var.strip('%')]:
+                        tpl[key] = tpl[key].replace(var, variables[var.strip('%')])
+                    else:
+                        log.warning('Failed to find a value for the {0} parameter.'
+                                    ' The check might not be configured properly.'.format(key))
+                        tpl[key].replace(var, '')
+        config[1] = config[1]
+        return config
+
+
+class SDDockerBackend(ServiceDiscoveryBackend):
+    """Docker-based service discovery"""
+
+    def __init__(self, agentConfig):
+        self.docker_client = get_docker_client()
+        self.VAR_MAPPING = {
+            'host': self._get_host,
+            'port': self._get_port,
+        }
+        ServiceDiscoveryBackend.__init__(self, agentConfig)
+
+    def _get_host(self, container_inspect):
+        """Extract the host IP from a docker inspect object."""
+        ip_addr = container_inspect['NetworkSettings']['IPAddress']
+        if not ip_addr:
+            # kubernetes case
+            c_id = container_inspect.get('Id')
+            task_id = self.docker_client.exec_create(c_id, 'hostname -I').get('Id')
```

Comment:
Especially that in the future we'll move to a "read only" role in the docker api when it will be implemented, so we won't be able to run that.